File: multi-lang/netcore/BatchResult.cs
namespace QQZeng.Qzdb;

/// <summary>Per-row result carrying the three-state lookup semantics of find().</summary>
/// <param name="Info">Resolved GeoInfo on success; null when not found or on error.</param>
/// <param name="Error">The error (e.g. InvalidIp / Corrupted) when the lookup failed; null on success or not-found.</param>
/// <param name="Input">The original IP string this result corresponds to (null for single find); lets callers trace batch/stream results back to their input.</param>
public readonly record struct BatchResult(GeoInfo? Info, QzdbException? Error, string? Input = null)
{
    /// <summary>True when the lookup succeeded and produced a GeoInfo.</summary>
    public bool IsSuccess => Error == null && Info != null;
    /// <summary>True when the lookup completed with no error and no result (a genuine miss).</summary>
    public bool IsNotFound => Error == null && Info == null;
    /// <summary>True when the lookup failed (e.g. InvalidIp / Corrupted); <see cref="Error"/> carries the detail.</summary>
    public bool HasError => Error != null;
}

File: qzdb/java/src/main/java/com/qqzeng/qzdb/BatchResult.java
package com.qqzeng.qzdb;

import java.util.Optional;

/**
 * 批量查询单条结果包装 (保留 命中 / 未找到 / 输入错误 完整精度)
 *
 * @param input  输入的原始 IP 文本
 * @param result 查询结果，若未找到则为 Optional.empty()
 * @param error  若输入格式错误或发生故障则非 null，正常或未找到时为 null
 */
public record BatchResult(
        String input,
        Optional<GeoInfo> result,
        QzdbException error
) {
    /**
     * 是否查询成功且命中记录
     */
    public boolean isSuccess() {
        return error == null && result.isPresent();
    }

    /**
     * 是否为有效 IP 但未找到记录
     */
    public boolean isNotFound() {
        return error == null && result.isEmpty();
    }

    /**
     * 是否发生输入格式错误或底层故障
     */
    public boolean hasError() {
        return error != null;
    }
}
